### PR TITLE
Add PixelCNN++ example

### PR DESCRIPTION
**Pull request by j-towns**
**EDIT:** This should be ready to merge, as you can see I haven't done everything on my original TODO list but what we do have now should be enough to land and iterate on.

TODO:
 - [x] Finish README
 - [x] Train to convergence
 - [x] Maybe simplify train.py
 - [x] Fix failing tests
 - [ ] Test a training iteration
 - [ ] Test receptive field
 - [x] Add checkpointing
 - [ ] Add naïve sampling

### CODE REVIEW

File: examples/pixelcnn/train.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+  metrics = jax.lax.pmean(metrics, 'batch')
+  return metrics
+
+
+def train_step(optimizer, state, batch, prng_key, learning_rate_fn):
+  """Perform a single training step."""
+  def loss_fn(model):
+    """loss function used for training."""
+    # TODO(j-towns): do we need state for this model?
+    with flax.nn.stateful(state) as new_state:
```
Comment (MostafaDehghani) [inline examples/pixelcnn/train.py:126]: ditto.

File: examples/pixelcnn/train.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+
+  metrics = {'loss': jax.lax.pmean(loss, 'batch')}
+  metrics['learning_rate'] = lr
+  return new_optimizer, new_state, metrics
+
+
+def eval_step(model, state, batch):
+  state = jax.lax.pmean(state, 'batch')
+  with flax.nn.stateful(state, mutable=False):
```
Comment (MostafaDehghani) [inline examples/pixelcnn/train.py:146]: ditto.

File: examples/pixelcnn/train.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+
+  base_learning_rate = learning_rate
+
+  # Create the model using data-dependent initialization. Don't shard the init
+  # batch.
+  assert init_batch_size <= batch_size
+  init_batch = next(train_iter)['image']._numpy()[:init_batch_size]
+  model, state = create_model(rng, init_batch, model_def)
+  state = jax_utils.replicate(state)
+  optimizer = create_optimizer(model, base_learning_rate)
```
Comment (MostafaDehghani) [inline examples/pixelcnn/train.py:199]: Consider adding checkpointing. It's actually pretty easy. Check the ImageNet example for that.

File: examples/pixelcnn/pixelcnn.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+  """
+  inputs = np.asarray(inputs, kwargs.get('dtype', np.float32))
+  k_h, k_w = kernel_size
+  out_h, out_w = onp.multiply(strides, inputs.shape[1:3])
+  return ConvTranspose(inputs, features, kernel_size, strides, **kwargs)[
+      :, :out_h, :out_w]
+
+
+# Resnet modules
+def make_res(conv_module):
```
Comment (MostafaDehghani) [inline examples/pixelcnn/pixelcnn.py:269]: Maybe call it make_gated_resnet?

File: examples/pixelcnn/model_test.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+from jax import random
+import jax.numpy as np
+
+from flax import nn
+
+import pixelcnn
+
+
+class ModelTest(absltest.TestCase):
```
Comment (MostafaDehghani) [inline examples/pixelcnn/model_test.py:17]: Super nice!

Comment (MostafaDehghani) [reply examples/pixelcnn/model_test.py:17]: Given all these masked convolutions, would be awesome to also include a test to check the receptive field of the model. 
Similar to what we have here: https://github.com/google/flax/blob/master/tests/nn_attention_test.py#L146

File: examples/pixelcnn/train.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    'n_logistic_mix', default=5,
+    help=('Number of components in the output distribution.'))
+
+flags.DEFINE_string(
+    'model_dir', default='./model_data',
+    help=('Directory to store model data'))
+
+
+def create_model(prng_key, example_images, model_def):
+  with flax.nn.stateful() as init_state:
```
[the remark this answers is not in the document]
Comment (MostafaDehghani) [reply examples/pixelcnn/train.py:88]: getting rid of this, you can remove all the state-related code and it's gonna be much cleaner.

File: examples/pixelcnn/train.py (hunk 3 above)
Comment (j-towns) [reply examples/pixelcnn/train.py:201]: Thanks, done.

**Comment by avital** [comment]
This LGTM (I mostly skimmed it).

It'd be also great to include generated samples in the README. I know you haven't yet committed your sampling code but could you keep a note to add that once sampling is ready?
